File: cores/go/internal/mapper/cnrom.go
// cnrom.go - Mapper 3 (CNROM). Port of cores/c/src/cnrom.c.
package mapper

func CnromCreate(m *Mapper, prg, chr []byte, mirroring Mirroring, hasBattery bool) int {
	s := &m.CNROM
	s.PrgSize = uint32(len(prg))
	if s.PrgSize > 0 {
		s.PrgRom = make([]byte, s.PrgSize)
		copy(s.PrgRom, prg)
	}
	if len(chr) == 0 {
		s.ChrSize = 8192
		s.Chr = make([]byte, s.ChrSize)
		s.ChrIsRAM = true
	} else {
		s.ChrSize = uint32(len(chr))
		s.Chr = make([]byte, s.ChrSize)
		copy(s.Chr, chr)
		s.ChrIsRAM = false
	}
	s.Mirroring = mirroring
	s.HasBattery = hasBattery
	return 0
}

func cnromReadPRG(m *Mapper, addr uint16) uint8 {
	s := &m.CNROM
	if addr < 0x8000 {
		return 0
	}
	local := uint32(addr - 0x8000)
	if s.PrgSize == 0 {
		return 0
	}
	return s.PrgRom[local%s.PrgSize]
}

func cnromWritePRG(m *Mapper, addr uint16, value uint8) {
	s := &m.CNROM
	if addr >= 0x8000 {
		s.ChrBank = value & 0x03
	}
}

func cnromReadCHR(m *Mapper, addr uint16) uint8 {
	s := &m.CNROM
	if s.ChrSize == 0 {
		return 0
	}
	bankSize := uint32(8192)
	if s.ChrSize < bankSize {
		bankSize = s.ChrSize
	}
	bank := uint32(s.ChrBank) % (s.ChrSize / bankSize)
	idx := bank*bankSize + (uint32(addr) % bankSize)
	if idx >= s.ChrSize {
		return 0
	}
	return s.Chr[idx]
}

func cnromWriteCHR(m *Mapper, addr uint16, value uint8) {
	s := &m.CNROM
	if !s.ChrIsRAM {
		return
	}
	if s.ChrSize == 0 {
		return
	}
	s.Chr[uint32(addr)%s.ChrSize] = value
}

func cnromSaveState(m *Mapper, buf []byte) int {
	s := &m.CNROM
	total := 4 + int(s.PrgSize) + 4 + 1 + 1 + 1 + 1
	if s.ChrIsRAM {
		total += int(s.ChrSize)
	}
	if buf == nil {
		return total
	}
	off := 0
	putU32(buf[off:], s.PrgSize); off += 4
	if s.PrgSize > 0 {
		copy(buf[off:], s.PrgRom); off += int(s.PrgSize)
	}
	putU32(buf[off:], s.ChrSize); off += 4
	buf[off] = boolByte(s.ChrIsRAM); off++
	buf[off] = uint8(s.Mirroring); off++
	buf[off] = boolByte(s.HasBattery); off++
	buf[off] = s.ChrBank; off++
	if s.ChrIsRAM && s.ChrSize > 0 {
		copy(buf[off:], s.Chr); off += int(s.ChrSize)
	}
	return off
}

func cnromLoadState(m *Mapper, buf []byte) bool {
	s := &m.CNROM
	if len(buf) < 4 {
		return false
	}
	off := 0
	prgSize := getU32(buf[off:]); off += 4
	if prgSize != s.PrgSize {
		return false
	}
	if s.PrgSize > 0 {
		copy(s.PrgRom, buf[off:off+int(s.PrgSize)]); off += int(s.PrgSize)
	}
	if len(buf) < off+4+4 {
		return false
	}
	chrSize := getU32(buf[off:]); off += 4
	if chrSize != s.ChrSize {
		return false
	}
	s.ChrIsRAM = buf[off] != 0; off++
	s.Mirroring = Mirroring(buf[off]); off++
	s.HasBattery = buf[off] != 0; off++
	s.ChrBank = buf[off]; off++
	if s.ChrIsRAM && s.ChrSize > 0 {
		if len(buf) < off+int(s.ChrSize) {
			return false
		}
		copy(s.Chr, buf[off:off+int(s.ChrSize)]); off += int(s.ChrSize)
	}
	return true
}

File: cores/c/src/cnrom.c
/*
 * cnrom.c - Mapper 3 (CNROM). Port of src/mappers/cnrom.rs to C (M4.4).
 *
 * CHR counterpart to UxROM: no PRG bank switching (PRG-ROM mapped linearly
 * across $8000-$FFFF, with 16 KB mirroring for single-bank carts), but CHR
 * is bank-switched in 8 KB windows. The bank register is written anywhere
 * in $8000-$FFFF (low 2 bits select 1 of 4 CHR banks). No PRG-RAM, no IRQ.
 *
 * See: https://www.nesdev.org/wiki/CNROM
 */
#include "mapper.h"
#include "cartridge.h"
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define PRG_BANK_SIZE 16384u
#define CHR_BANK_SIZE 8192u

typedef struct CnromState {
    uint8_t* prg_rom;
    uint32_t prg_size;
    uint8_t* chr;
    uint32_t chr_size;
    bool     chr_is_ram;
    Mirroring mirroring;
    bool     has_battery;
    uint8_t  chr_bank; /* currently selected 8 KB CHR bank. */
} CnromState;

static uint32_t cnrom_chr_bank_count(const CnromState* s) {
    uint32_t n = s->chr_size / CHR_BANK_SIZE;
    return n > 0u ? n : 1u;
}

static uint32_t cnrom_prg_index(const CnromState* s, uint16_t addr) {
    uint32_t local = (uint32_t)(addr - 0x8000u);
    uint32_t bank = s->prg_size;
    if (bank == 0u) {
        return 0u;
    }
    return local % bank;
}

static uint8_t cnrom_read_prg(const void* state, uint16_t addr) {
    const CnromState* s = (const CnromState*)state;
    if (addr < 0x8000u) {
        return 0x00u;
    }
    uint32_t idx = cnrom_prg_index(s, addr);
    if (idx >= s->prg_size) {
        return 0x00u;
    }
    return s->prg_rom[idx];
}

static void cnrom_write_prg(void* state, uint16_t addr, uint8_t value) {
    CnromState* s = (CnromState*)state;
    if (addr < 0x8000u) {
        return;
    }
    s->chr_bank = value;
}

static uint8_t cnrom_read_chr(const void* state, uint16_t addr) {
    const CnromState* s = (const CnromState*)state;
    uint32_t count = cnrom_chr_bank_count(s);
    uint32_t bank = (uint32_t)s->chr_bank % count;
    uint32_t idx = bank * CHR_BANK_SIZE + ((uint32_t)addr & (CHR_BANK_SIZE - 1u));
    if (idx >= s->chr_size) {
        return 0x00u;
    }
    return s->chr[idx];
}

static void cnrom_write_chr(void* state, uint16_t addr, uint8_t value) {
    CnromState* s = (CnromState*)state;
    if (!s->chr_is_ram) {
        return;
    }
    uint32_t count = cnrom_chr_bank_count(s);
    uint32_t bank = (uint32_t)s->chr_bank % count;
    uint32_t idx = bank * CHR_BANK_SIZE + ((uint32_t)addr & (CHR_BANK_SIZE - 1u));
    if (idx < s->chr_size) {
        s->chr[idx] = value;
    }
}

static Mirroring cnrom_mirror_mode(const void* state) {
    return ((const CnromState*)state)->mirroring;
}

static bool cnrom_chr_is_ram(const void* state) {
    return ((const CnromState*)state)->chr_is_ram;
}

static bool cnrom_has_battery(const void* state) {
    return ((const CnromState*)state)->has_battery;
}

static void cnrom_destroy(void* state) {
    CnromState* s = (CnromState*)state;
    if (s) {
        free(s->prg_rom);
        free(s->chr);
        free(s);
    }
}

static const MapperVTable CNROM_VTABLE = {
    cnrom_read_prg, NULL, cnrom_write_prg,
    cnrom_read_chr, NULL, cnrom_write_chr,
    cnrom_mirror_mode, cnrom_chr_is_ram, cnrom_has_battery,
    NULL, NULL, NULL, NULL, NULL, cnrom_destroy
};

int cnrom_create(const uint8_t* prg, uint32_t prg_size,
                 const uint8_t* chr, uint32_t chr_size,
                 Mirroring mirroring, bool has_battery, Mapper* out) {
    CnromState* s = (CnromState*)malloc(sizeof(CnromState));
    if (!s) {
        return 1;
    }
    s->prg_size = prg_size;
    s->prg_rom = NULL;
    if (prg_size > 0u) {
        s->prg_rom = (uint8_t*)malloc(prg_size);
        if (!s->prg_rom) {
            free(s);
            return 1;
        }
        memcpy(s->prg_rom, prg, prg_size);
    }
    /* CHR: 8 KB RAM if no CHR-ROM, else copy CHR-ROM (may be > 8 KB). */
    if (chr_size == 0u) {
        s->chr_size = CHR_BANK_SIZE;
        s->chr = (uint8_t*)malloc(CHR_BANK_SIZE);
        if (!s->chr) {
            free(s->prg_rom);
            free(s);
            return 1;
        }
        memset(s->chr, 0, CHR_BANK_SIZE);
        s->chr_is_ram = true;
    } else {
        s->chr_size = chr_size;
        s->chr = (uint8_t*)malloc(chr_size);
        if (!s->chr) {
            free(s->prg_rom);
            free(s);
            return 1;
        }
        memcpy(s->chr, chr, chr_size);
        s->chr_is_ram = false;
    }
    s->mirroring = mirroring;
    s->has_battery = has_battery;
    s->chr_bank = 0u;
    out->vt = &CNROM_VTABLE;
    out->state = s;
    out->mapper_num = 3u;
    return 0;
}
